src/node_crypto.cc

```diff
@@ -5771,6 +5771,61 @@ void ExportChallenge(const FunctionCallbackInfo<Value>& args) {
   args.GetReturnValue().Set(outString);
 }
 
+void TimingSafeEqual(const FunctionCallbackInfo<Value>& args) {
+  // An implementation of the double HMAC pattern described here:
+  // https://www.nccgroup.trust/us/about-us/newsroom-and-events/blog/2011/
+  // february/double-hmac-verification/
+
+  // Instead of attempting a comparison that always takes the same amount of
+  // time (which is difficult to do properly due to unpredictable compiler
+  // optimizations), this algorithm blinds the timing channel by computing an
+  // HMAC of each buffer with the same random key. The resulting HMAC digests
+  // can then be directly compared safely, because there is no correlation
+  // between the comparison time for these digests and the similarity between
+  // the original buffers.
+  Environment* env = Environment::GetCurrent(args);
+
+  THROW_AND_RETURN_IF_NOT_BUFFER(args[0], "First argument");
+  THROW_AND_RETURN_IF_NOT_BUFFER(args[1], "Second argument");
+
+  char* buf1 = Buffer::Data(args[0]), *buf2 = Buffer::Data(args[1]);
+
+  size_t buf_length = Buffer::Length(args[0]);
+  if (buf_length != Buffer::Length(args[1])) {
+    return env->ThrowTypeError("Input buffers must have the same length");
+  }
+
+  Local<Object> obj = env->NewInternalFieldObject();
+
+  const int64_t HMAC_KEY_SIZE = 32;
+  RandomBytesRequest* req = new RandomBytesRequest(env, obj, HMAC_KEY_SIZE);
+
+  Local<Value> argv[2];
+  RandomBytesWork(req->work_req());
+  RandomBytesCheck(req, argv);
+
+  Hmac* hmac1 = new Hmac(env, obj), *hmac2 = new Hmac(env, obj);
+
+  unsigned char* digest1 = nullptr, *digest2 = nullptr;
+  unsigned int digest_len1 = 0, digest_len2 = 0;
+
+  hmac1->HmacInit("sha256", Buffer::Data(argv[1]), HMAC_KEY_SIZE);
+  hmac1->HmacUpdate(buf1, buf_length);
+  hmac1->HmacDigest(&digest1, &digest_len1);
+
+  hmac2->HmacInit("sha256", Buffer::Data(argv[1]), HMAC_KEY_SIZE);
+  hmac2->HmacUpdate(buf2, buf_length);
+  hmac2->HmacDigest(&digest2, &digest_len2);
```

Comment:
Calling RandomBytes{Work,Check} and Hmac{Init,Update,Digest} feels very roundabout.  Why don't you call the relevant openssl APIs directly?